Consider ThirdPartyCookieBlockingMode::AllExceptPartitioned when deciding if cookie access should be granted via the Storage Access API

sysrqb opened the pull request:
#### 845f0a548125461da779d734a84b15498a4538cb
<pre>
Consider ThirdPartyCookieBlockingMode::AllExceptPartitioned when deciding if cookie access should be granted via the Storage Access API
<a href="https://bugs.webkit.org/show_bug.cgi?id=286428">https://bugs.webkit.org/show_bug.cgi?id=286428</a>
<a href="https://rdar.apple.com/143508260">rdar://143508260</a>

Reviewed by John Wilander and Brent Fulgham.

When handling a requestStorageAccess request, we considered storage access as
blocked only if all third-party cookies were blocked. In theory, this is true,
because the site has access to partitioned cookies, but that misses the point
of requesting storage access where the site gets access to its first party
cookies. This patch extends the current logic so it considers
AllExceptPartitioned as not currently having cookie access - therefore requires
handling the request for storage access.

Covered by existing tests.

* Source/WebKit/NetworkProcess/Classifier/ResourceLoadStatisticsStore.cpp:
(WebKit::ResourceLoadStatisticsStore::areAllUnpartitionedThirdPartyCookiesBlockedUnder):
(WebKit::ResourceLoadStatisticsStore::cookieAccess):
(WebKit::ResourceLoadStatisticsStore::areAllThirdPartyCookiesBlockedUnder): Deleted.
* Source/WebKit/NetworkProcess/Classifier/ResourceLoadStatisticsStore.h:

Canonical link: <a href="https://commits.webkit.org/289450@main">https://commits.webkit.org/289450@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/11b00682ea5274a1fa5286a0d5303fef9a59e6d6

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/86948 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/6456 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/41348 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/91801 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/37733 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/6724 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/14573 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/67245 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/25007 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/89951 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/5187 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/78748 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/47567 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/4965 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/36856 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/75458 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/33998 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/93741 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/14157 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/10298 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/76048 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/14360 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/74601 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/75246 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/19580 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/18009 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/6988 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/13554 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/14125 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/19418 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/13921 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/17364 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/15652 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

brentfulgham commented on Source/WebKit/NetworkProcess/Classifier/ResourceLoadStatisticsStore.cpp:
```diff
@@ -2542,9 +2542,13 @@ void ResourceLoadStatisticsStore::clear(CompletionHandler<void()>&& completionHa
+bool ResourceLoadStatisticsStore::areAllUnpartitionedThirdPartyCookiesBlockedUnder(const TopFrameDomain& topFrameDomain)
 {
+#if HAVE(ALLOW_ONLY_PARTITIONED_COOKIES)
+    if (thirdPartyCookieBlockingMode() == ThirdPartyCookieBlockingMode::All || thirdPartyCookieBlockingMode() == ThirdPartyCookieBlockingMode::AllExceptPartitioned)
```
This seems fine, and obviously doesn't break cases that worked with "all" blocking mode. But how do we test partitioned cases?

sysrqb replied:
EWS doesn't have support for partitioned cookies right now, so this change won't actually be tested here. We have other internal tests that verify this change.

brentfulgham reviewed (approved):
Change itself looks fine, but would like to know more about how we ensure this continues to work as expected.

johnwilander reviewed (approved):
lgtm

Commit:
Consider ThirdPartyCookieBlockingMode::AllExceptPartitioned when deciding if cookie access should be granted via the Storage Access API
https://bugs.webkit.org/show_bug.cgi?id=286428
rdar://143508260

Reviewed by John Wilander and Brent Fulgham.

When handling a requestStorageAccess request, we considered storage access as
blocked only if all third-party cookies were blocked. In theory, this is true,
because the site has access to partitioned cookies, but that misses the point
of requesting storage access where the site gets access to its first party
cookies. This patch extends the current logic so it considers
AllExceptPartitioned as not currently having cookie access - therefore requires
handling the request for storage access.

Covered by existing tests.

* Source/WebKit/NetworkProcess/Classifier/ResourceLoadStatisticsStore.cpp:
(WebKit::ResourceLoadStatisticsStore::areAllUnpartitionedThirdPartyCookiesBlockedUnder):
(WebKit::ResourceLoadStatisticsStore::cookieAccess):
(WebKit::ResourceLoadStatisticsStore::areAllThirdPartyCookiesBlockedUnder): Deleted.
* Source/WebKit/NetworkProcess/Classifier/ResourceLoadStatisticsStore.h:

Canonical link: https://commits.webkit.org/289450@main
